plasma_framework/contracts/src/framework/utils/ExitPriority.sol

```diff
@@ -3,21 +3,32 @@ pragma solidity ^0.5.0;
 import "../../utils/TxPosLib.sol";
 
 library ExitPriority {
+
     /**
-     * @dev formula of priority is as folllowed: (exitableAt || txPos || nonce).
-     * The first 64 bit for exitableAt, following 128 bits of txPos and then 64 bits of nonce.
-     * The combinition of 'exitableAt' and 'txPos' is the priority for Plasma M(ore)VP protocol.
-     * 'exitableAt' only provide granuality of block, thus add 'txPos' to provide priority of transaction.
+     * @dev Given a utxo position and a unique ID, returns an exit priority.
+     * The combination of 'exitableAt' and 'txPos' is the priority for Plasma M(ore)VP protocol.
+     * 'exitableAt' only provide granularity of block, thus add 'txPos' to provide priority of transaction.
+     * @param exitId Unique exit identifier.
+     * @return An exit priority.
+     *   Anatomy of returned value, most significant bits first
+     *   42 bits - timestamp (exitable_at); we can represent dates until 2109 
+     *   54 bits - blknum * 10^9 + txindex; to represent all transactions for 10 years we need only 54 bits
```

Comment:
This is still wrong though. 
should be `blknum * 10^5 + txindex`, and of would have much longer life span than 10 years.

Reply:
it might be worthwhile to add a reference (or a quick dissection in here) towards a source of this calculation and the 10^5 part. I take it it comes from here:

https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/contracts/src/utils/TxPosLib.sol#L11

and here:

https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/contracts/src/utils/TxPosLib.sol#L11

Also note a related issue https://github.com/omisego/plasma-contracts/issues/254, so that we not forget to update those docs, in case this modification gets through.

I think it's very important to keep all those offsetting and bitwise operations as transparent and documented as possible.

Reply:
I made some more caution maths and it the truth appears to be yet another thing :joy:

### tx pos
Considering a UTXO_POS schema:
https://github.com/omisego/plasma-contracts/blob/c4aad906bb948bc59d1d7eda653a8219263d44e3/plasma_framework/contracts/src/utils/UtxoPosLib.sol#L6

We get than in each block there are at most `BLOCK_OFFSET / TX_OFFSET = 10^5` transactions, but we should also take into account `CHILD_BLOCK_INTERVAL` as each block number is multiplied by it (not counting deposits).
Right now `CHILD_BLOCK_INTERVAL` is set to `10^3`, and it determines number of deposits that can be made between the two consecutive blocks submitted be the operator (kudos to @pdobacz for pointing this out).

Therefore, we should reserve `10^8` per child chain block. Assuming that we commit a new child chain block in each root chain block (which happens every ~15 seconds), we must reserve `10^8 * 4 * 60 * 24 * 365` space a year (for simplicity let's don't take into account leap years). (please point out if my math is wrong there)

Having this said and having reserved 54-bits for tx pos, we will be out of space in ~85 years after the initial block - if we started now we would be out of space in ~2104.

### exitable timestamp
There is another bug with the timestamp: 42-bits will finish in year 141431 :joy:
Calculating the 2109, I took `2^42`, put it into https://www.epochconverter.com and received value 2109, but I did not notice that the smart website assumed the the value provided was in _milliseconds_ instead of seconds.

### Another solutions
I may just amend the comments, but there is also one more possible solution. Having done the math, we see that timestamp has ridiculously much space, on the contrary to tx pos; we can move some bits from timestamp to tx pos, in such a way that the years they exhaust are more "balanced". We could use 37-bits for timestamp and 59-bits for tx pos, in such case they will end in years 6328 and 4760 (2741 years since start) respectively.

We can also do one more thing: use  33-bits for timestamp and 55-bits for tx pos ( timestamp finishes in 2242 and tx pos has space to run for 171 years) and move the additional 8 bits to exit id and end up with 168-bits for exit id.

Reply:
> Right now CHILD_BLOCK_INTERVAL is set to 10^3

oh......I definitely missed this! Meanwhile, I think [Eth2.0 would make the 15 sec -> 6 sec](https://our.status.im/9-things-you-didnt-know-about-ethereum-2-0/), so would be half to 1/3 of our current calculation.

>  in milliseconds instead of seconds.

ah....I missed this whenever I was doing the calculation too @@ So yeah, it seems like it has more than enough bits. Let's re-balance this.

> 168-bits for exit id.

Though I really want the exitId to be longer so that it has at least full 160 bits for hash collision protection, but since there is no type `uint168`, I would prefer we stick to either `uint192` or `uint160`. Otherwise passing `uint192` with a real limit of `uint168` just make things more complex.

Reply:
> I would prefer we stick to either `uint192` or `uint160`.

Unfortunatelly we cannot use 192-bits since we cannot fit both timestamp and tx pos in only 64 bits.

Reply:
>BTW, all this complexity is what I think original ALD schema is better actually. It is just more easy to analyze relatively. And, either way we are approaching it is breaking the client anyway @@....

That's why I suggest just changing the documentation and explicitly clarify the numbers there.

The problem with the ALD schema is the overall cost - we store more bytes, we use nonces which we update each time which costs a loooot of gas.